core/trie/proof.go

```
@@ -0,0 +1,112 @@
+	if err != nil {
+		return nil, err
+	}
+	nodesExcludingLeaf := nodesToLeaf[:len(nodesToLeaf)-1]
+	proofNodes := make([]ProofNode, len(nodesExcludingLeaf))
+
+	getHash := func(key *Key) (*felt.Felt, error) {
+		keyFelt := key.Felt()
+		node, err := tri.GetNode(&keyFelt)
+		if err != nil {
+			return nil, err
+		}
+		return node.Hash(key, crypto.Pedersen), nil
+	}
+
+	// The spec requires Edge nodes. However Juno doesn't have edge nodes.
+	// if Len=250 and have left and right -> create a Binary
+	// if len=250 and only have one child -> create edge
+	// if len!=250, check if child-parent length = 1, if so ,create Binary
+	// if len!=250, check if child-parent length = 1, if not ,create Edge
+	for i, sNode := range nodesExcludingLeaf {
+		// sLeft := sNode.node.Left
+		// sRight := sNode.node.Right
+		nxtNode := nodesToLeaf[i+1]
+
+		if nxtNode.key.len-sNode.key.len > 1 { // split node into edge + child
+			edgePath := NewKey(sNode.key.len, sNode.key.bitset[:])
+			edgePath.RemoveLastBit() // Todo: make sure we remove it from the correct side
+			edgePathFelt := edgePath.Felt()
+
+			childKeyFelt := sNode.key.Felt()
+
+			proofNodes[i] = ProofNode{
+				Edge: &Edge{
+					Path:  &edgePathFelt,
+					Child: &childKeyFelt,
+					// Value: value, // Todo: ??
+				},
+			}
+		}
+		leftHash, err := getHash(sNode.node.Left)
+		if err != nil {
+			return nil, err
+		}
+		rightHash, err := getHash(sNode.node.Right)
+		if err != nil {
+			return nil, err
+		}
+		proofNodes[i] = ProofNode{
+			Binary: &Binary{
+				LeftHash:  leftHash,
+				RightHash: rightHash,
+			},
+		}
+	}
+
+	return proofNodes, nil
+}
+
+// verifyProof checks if `leafPath` leads from `root` to `leafHash` along the `proofNodes`
+// https://github.com/eqlabs/pathfinder/blob/main/crates/merkle-tree/src/tree.rs#L2006
+// func VerifyProof(root *felt.Felt, leafPath *Key, leafHash felt.Felt, proofNodes []ProofNode, hashFunc hashFunc) error {
+// 	expectedHash := root
+
+// 	for i, pNode := range proofNodes {
```

Keep in mind, when using a proof for a range of leaf, you'll need to make sure that it fail when a leaf in the middle of the tree is missing. Which is a bit more complicated.